### ISSUE TITLE
Question: should no-floating-promises consider queueMicrotask() as well?

### ISSUE THREAD

**Issue by mathiasbynens** [typescript-eslint/typescript-eslint#4500]
### Description

[`no-floating-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-floating-promises.md) doesn’t do anything special w.r.t. `queueMicrotask(fn)`, despite it being a cross-platform way to execute a given `function` on the next micro-tick. It’s similar to `Promise.resolve(fn)` except it doesn’t have a return value. So, it could arguably be considered as a more self-descriptive way to “explicitly mark a promise as intentionally not awaited”, alongside the already-supported `void Promise.resolve(fn)`. WDYT?

### Fail

N/A

### Pass

N/A

**Comment by JoshuaKGoldberg** [comment]
Hmm, good question. It doesn't handle `setInterval` or `setTimeout` either for that matter. Is there a code snippet you're thinking about this in the context of @mathiasbynens?

I'll note that this isn't a base rule extension as `no-floating-promises` doesn't exist in ESLint core. It's purely a Typescript ESLint rule.

**Comment by mathiasbynens** [comment]
I think `queueMicrotask` makes sense to consider since it schedules a microtask, just like `Promise.resolve()` does.

`setInterval` or `setTimeout` however don’t schedule microtasks, so I don’t see how they might be relevant for this rule related to promises. Can you elaborate?

**Comment by mathiasbynens** [comment]
Here’s a code example: https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/3427063

DevTools had this in its code base, using `void` to mark the promise as explicitly not `await`ed:

```js
void Promise.resolve().then(progress.done.bind(progress));
```

I claim that the following rewrite is just as clear (and arguably more descriptive / more concise):

```js
queueMicrotask(progress.done.bind(progress));
```

**Comment by mathiasbynens** [comment]
1. `no-floating-promises` attempts to find floating promises that e.g. aren’t being `await`ed (or handled in some other way). `queueMicrotask` implicitly creates such a promise (similar to how `Promise.resolve()` does, except `queueMicrotask` doesn’t return that promise). Hence my thinking that perhaps this plugin should detect its usage.

2. But OTOH, since that’s exactly what `queueMicrotask` is meant to do, its use could reasonably be treated as a signal that the implicit promise it creates is “intentionally not `await`ed”, similar to `void` when [the `ignoreVoid` option](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-floating-promises.md#ignorevoid) is set. Since the current plugin ignores `queueMicrotask` this already “works”, but depending on whether you decide to do anything w.r.t. point 1 above, it might need some special handling.

**Comment by bradzacher** [comment]
Sorry - I understand the similar behaviour. Instead what I'm asking is - what is floating after the `queueMicroTask` call? What should the rule be doing?

As I understand it - nothing is left floating because the api has no return value. There's no promise to call `.catch` or `.finally` on, or to `await`. 

If you're talking about passing a promise into `queueMicroTask`, then that won't be flagged by the rule. A promise passed to a function is not floating. A promise is only floating if it's the result of an expression statement.

> Hence my thinking that perhaps this plugin should detect its usage.

What should the rule detect, exactly? What should the rule report? How would someone fix it? What shouldn't it detect?

**Comment by mathiasbynens** [comment]
> What should the rule detect, exactly? What should the rule report? How would someone fix it? What shouldn't it detect?

I'm not sure, that's why I asked :) It's fine if the answer is it shouldn't treat this case specially. I just wanted to double-check it was a conscious choice and not some oversight.

I mostly felt like `queueMicrotask` is a more descriptive way of what the rule currently recommends doing with `void`,  in the specific case where `Promise.resolve().then(fn)` is used. But like I said this already "works" in the sense that the rule at least doesn't complain about this, so perhaps there's nothing to be done here.

**Comment by bradzacher** [comment]
Yeah it's fully intentional that `queueMicrotask` is ignored by the rule.
The rule specifically inspects expression statements and interrogates them if and only if their type is promise-like.
`queueMicrotask` returns nothing - so its expression statements are ignored fully.

As for `queueMicrotask` vs `void` - they both have very different usages!
`queueMicrotask` specifically accepts a callback function, not a promise, and it schedules it for execution "at the end of the event loop".
OTOH the `void` operator accepts anything and always returns `undefined`. It does not change how or when the work is done.

The reason `void` is suggested by the rule is really for convention and convenience. There are very many times that people intentionally kick off async work and do not `await` / `.catch` it. As a way of signifying that the promise was intentionally floating, use of the keyword `void` emerged as a standard - I assume because using a keyword created a nice symmetry with `await`. This pattern existed before the rule, so it was added by popular demand.

Additionally people liked the conciseness of
```ts
void myAsyncTask()
```
as opposed to
```ts
// eslint-disable-next-line @typescript-eslint/no-floating-promises
myAsyncTask()
```
Adding it to the rule also meant that you can then also enforce the inverse (i.e. that people use the disable comment instead of the `void` operator).

---

I would definitely agree that `queueMicrotask` is better that `Promise.resolve().then` - but that would be the domain of a separate lint rule!

**Comment by mathiasbynens** [comment]
> There are very many times that people intentionally kick off async work and do not `await` / `.catch` it.

I get all that. All I'm saying is that `queueMicrotask` is one of those times. (`setTimeout`/`setInterval` aren't, since they don't schedule microtasks.)

> Yeah it's fully intentional that `queueMicrotask` is ignored by the rule.

Alright, thank you!
